# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: layout/painting/RetainedDisplayListBuilder.cpp
# Commit: 984c175f86f9
# Full Hash: 984c175f86f98a22d6a56e40773f5e3f773a004a
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-05-02 04:38:13
# Description:
#   Bug 1547986 - Don't early return in PreProcessDisplayList if we need to re-link the display list on exit. r=miko
#   
#   This early return is just an optimization to prevent the DAG from becoming too complex, and if we're keeping the list linked, then we know it won't be getting more complex on the current paint.
#   Future paints that actually modify the list will still take this path.
#   
# ==============================================================================

diff -r 6487aeb66175 -r 984c175f86f9 layout/painting/RetainedDisplayListBuilder.cpp
--- a/layout/painting/RetainedDisplayListBuilder.cpp	Wed May 01 17:11:19 2019 +0000
+++ b/layout/painting/RetainedDisplayListBuilder.cpp	Wed May 01 17:34:45 2019 +0000
@@ -130,7 +130,7 @@
   // list build if we hit them.
   static const uint32_t kMaxEdgeRatio = 5;
   const bool initializeDAG = !aList->mDAG.Length();
-  if (!initializeDAG && aList->mDAG.mDirectPredecessorList.Length() >
+  if (!aKeepLinked && !initializeDAG && aList->mDAG.mDirectPredecessorList.Length() >
                             (aList->mDAG.mNodesInfo.Length() * kMaxEdgeRatio)) {
     return false;
   }
@@ -227,6 +227,7 @@
       if (!PreProcessDisplayList(
               item->GetChildren(), SelectAGRForFrame(f, aAGR), aUpdated,
               item->GetPerFrameKey(), aNestingDepth + 1, keepLinked)) {
+        MOZ_RELEASE_ASSERT(!aKeepLinked, "Can't early return since we need to move the out list back");
         return false;
       }
     }
